core/pkg/resourcehandler/remotegitutils.go

```diff
@@ -18,9 +38,26 @@ func cloneRepo(gitURL giturl.IGitURL) (string, error) {
 		return "", fmt.Errorf("failed to create temporary directory: %w", err)
 	}
 
-	// Clone option
+	// Get the URL to clone
 	cloneURL := gitURL.GetHttpCloneURL()
-	cloneOpts := git.CloneOptions{URL: cloneURL}
+
+	isGitRepoPublic := isGitRepoPublic(cloneURL)
+
+	// Declare the authentication variable required for cloneOptions
+	var auth transport.AuthMethod
+
+	if isGitRepoPublic {
+		// No authentication needed if repository is public
+		auth = nil
+	} else {
+		auth = &http.BasicAuth{
+			Username: "anything Except Empty String",
+			Password: os.Getenv("GITHUB_TOKEN"),
```

Comment:
We should use the `giturl.IGitURL` interface for loading the token.
We might need to add a `GetToken` method to the interface

Reply:
* There is another interface called `giturl.IGitAPI` which inherits `giturl.IGitURL` has some additional methods including `GetToken`. 
* I tested the code using that interface (only small changes required) and the code is working fine. Shall we proceed with that?

Reply:
* There is another interface called `giturl.IGitAPI` which inherits `giturl.IGitURL` has some additional methods including `GetToken`.
* I tested the code using that interface (only small changes required) and the code is working fine. Shall we proceed with that?

Reply:
I think it is not a secure way to authenticate a user by setting their token in the environment.
I think one switch is also available like -p by using that flag users can pass their token in script or it prompts input for the token,
if someone put their token in their os environment they can be exploited.